File: examples/c/fft_scan.c
/*
 * fft_scan.c — FFT 주파수 분석
 * Build: make examples
 * Run  : ./examples/c/fft_scan [ip] [port] [sample_rate] [channel_mask] [samples] [--csv]
 *        sample_rate:  0=64K 1=128K 2=170K 3=256K 4=512K  (기본 0)
 *        channel_mask: 0x01~0x0F                           (기본 0x0F)
 *        samples:      수집 샘플 수, 2의 거듭제곱 권장       (기본: sample_rate ≈ 1초 분량)
 */
#include "prismlib.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>
#include <unistd.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

/* ── 상수 ────────────────────────────────────────────────────────────────── */
static const int   SR_HZ[]        = {64000, 128000, 170000, 256000, 512000};
static const char *SR_NAME[]      = {"64K",  "128K",  "170K",  "256K",  "512K"};
/* 각 샘플레이트의 ≈1초 분량, 2의 거듭제곱 */
static const int   SR_DEFAULT_N[] = {65536, 131072, 131072, 262144, 524288};

#define N_HARMONICS  5
#define MIN_FREQ_HZ  5.0

/* ── Cooley-Tukey radix-2 DIT in-place FFT ──────────────────────────────── */
static void fft_inplace(double *re, double *im, int n)
{
    int i, j = 0;
    for (i = 1; i < n; i++) {
        int bit = n >> 1;
        for (; j & bit; bit >>= 1) j ^= bit;
        j ^= bit;
        if (i < j) {
            double tr = re[i]; re[i] = re[j]; re[j] = tr;
            double ti = im[i]; im[i] = im[j]; im[j] = ti;
        }
    }
    int len;
    for (len = 2; len <= n; len <<= 1) {
        double ang = -2.0 * M_PI / len;
        double wre = cos(ang), wim = sin(ang);
        for (i = 0; i < n; i += len) {
            double cr = 1.0, ci = 0.0;
            for (j = 0; j < len / 2; j++) {
                double ure = re[i+j],        uim = im[i+j];
                double k   = re[i+j+len/2];
                double l   = im[i+j+len/2];
                double vre = k * cr - l * ci, vim = k * ci + l * cr;
                re[i+j]       = ure + vre;  im[i+j]       = uim + vim;
                re[i+j+len/2] = ure - vre;  im[i+j+len/2] = uim - vim;
                double nr = cr * wre - ci * wim;
                ci = cr * wim + ci * wre;
                cr = nr;
            }
        }
    }
}

/* ── 단측 진폭 스펙트럼 ─────────────────────────────────────────────────── */
static double *compute_spectrum(const double *signal, int n, double fs,
                                double **freqs_out)
{
    int bins = n / 2 + 1;
    double *re   = (double *)malloc(n * sizeof(double));
    double *im   = (double *)calloc(n, sizeof(double));
    double *amp  = (double *)malloc(bins * sizeof(double));
    double *freq = (double *)malloc(bins * sizeof(double));
    if (!re || !im || !amp || !freq) {
        free(re); free(im); free(amp); free(freq);
        return NULL;
    }

    double win_sum = 0.0;
    int i;
    for (i = 0; i < n; i++) {
        double w = 0.5 * (1.0 - cos(2.0 * M_PI * i / (n - 1)));
        re[i]    = signal[i] * w;
        win_sum += w;
    }
    double gain = win_sum / n;

    fft_inplace(re, im, n);

    for (i = 0; i < bins; i++) {
        amp[i]  = sqrt(re[i]*re[i] + im[i]*im[i]) / (n * gain);
        if (i > 0 && i < bins - 1) amp[i] *= 2.0;
        freq[i] = (double)i * fs / n;
    }

    free(re);
    free(im);
    *freqs_out = freq;
    return amp;
}

/* ── 기본파 + 고조파 피크 탐색 ──────────────────────────────────────────── */
static int find_peaks(const double *freq, const double *amp, int bins,
                      double *peak_freq, double *peak_amp, int n)
{
    double freq_res = bins > 1 ? freq[1] : 1.0;
    int    min_bin  = (int)ceil(MIN_FREQ_HZ / freq_res);
    if (min_bin < 1) min_bin = 1;

    int pk = min_bin, i;
    for (i = min_bin + 1; i < bins; i++)
        if (amp[i] > amp[pk]) pk = i;
    double fund_hz = freq[pk];

    int count = 0, h;
    for (h = 1; h <= n; h++) {
        int tbin = (int)round(fund_hz * h / freq_res);
        if (tbin >= bins) break;
        int lo   = (tbin - 2 > min_bin) ? tbin - 2 : min_bin;
        int hi   = (tbin + 2 < bins)    ? tbin + 2 : bins - 1;
        int best = lo;
        for (i = lo + 1; i <= hi; i++)
            if (amp[i] > amp[best]) best = i;
        peak_freq[count] = freq[best];
        peak_amp[count]  = amp[best];
        count++;
    }
    return count;
}

/* ── scan 정지 대기 ─────────────────────────────────────────────────────── */
static void wait_scan_stopped(prismlib_t *dev)
{
    uint16_t st; uint32_t av;
    int i;
    for (i = 0; i < 5000; i++) {
        prismlib_scan_status(dev, &st, &av);
        if (!(st & STATUS_RUNNING)) return;
        volatile int j = 0; while (j++ < 50000);
    }
}

/* ── main ────────────────────────────────────────────────────────────────── */
int main(int argc, char *argv[])
{
    const char *ip      = (argc > 1) ? argv[1] : "192.168.7.1";
    int         port    = (argc > 2) ? atoi(argv[2]) : 7777;
    int         sr      = (argc > 3) ? atoi(argv[3]) : 0;
    int         ch_mask = (argc > 4) ? (int)strtol(argv[4], NULL, 0) : 0x0F;
    int         save_csv = 0, i;
    for (i = 1; i < argc; i++)
        if (strcmp(argv[i], "--csv") == 0) save_csv = 1;

    if (sr < 0 || sr > 4) {
        fprintf(stderr, "sample_rate: 0=64K 1=128K 2=170K 3=256K 4=512K\n");
        return 1;
    }
    if (ch_mask < 0x01 || ch_mask > 0x0F) {
        fprintf(stderr, "channel_mask: 0x01~0x0F\n");
        return 1;
    }

    int n_samples = (argc > 5) ? atoi(argv[5]) : SR_DEFAULT_N[sr];

    int active[4], n_ch = 0;
    for (i = 0; i < 4; i++)
        if (ch_mask & (1 << i)) active[n_ch++] = i;

    double fs = (double)SR_HZ[sr];

    printf("Connecting to %s:%d ...\n", ip, port);
    prismlib_t *dev = prismlib_open(ip, (uint16_t)port);
    if (!dev) { fprintf(stderr, "prismlib_open failed\n"); return 1; }

    prismlib_sampleRate_write(dev, (PrismSampleRate_e)sr);

    {
        char ch_str[32] = {0};
        for (i = 0; i < n_ch; i++) {
            char tmp[16];
            snprintf(tmp, sizeof(tmp), "%sch%d", i ? "+" : "", active[i] + 1);
            strncat(ch_str, tmp, sizeof(ch_str) - strlen(ch_str) - 1);
        }
        printf("Sample rate  : %s (%d Hz)\n", SR_NAME[sr], SR_HZ[sr]);
        printf("Channels     : %s\n", ch_str);
        printf("Samples/ch   : %d  (freq resolution: %.3f Hz)\n",
               n_samples, fs / n_samples);
    }

    int use_iepe = 0;
    printf("IEPE를 활성화하시겠습니까? [y/n]: ");
    fflush(stdout);
    char ans[8] = {0};
    if (fgets(ans, sizeof(ans), stdin) && (ans[0] == 'y' || ans[0] == 'Y')) {
        use_iepe = 1;
        for (i = 0; i < n_ch; i++)
            prismlib_iepe_write(dev, (uint8_t)active[i], 1);
        printf("IEPE 안정화 대기 중 (2초)...");
        fflush(stdout);
        sleep(2);
        printf(" 완료\n");
    }

    printf("Collecting ...\n");
    fflush(stdout);

    /* ── 스캔 ────────────────────────────────────────────────────────────── */
    uint32_t buf_size = (uint32_t)(n_samples * n_ch);
    double  *buf      = (double *)malloc(buf_size * sizeof(double));
    if (!buf) {
        if (use_iepe)
            for (i = 0; i < n_ch; i++)
                prismlib_iepe_write(dev, (uint8_t)active[i], 0);
        prismlib_close(dev);
        return 1;
    }

    uint16_t status = 0;
    uint32_t got    = 0;
    prismlib_scan_start(dev, (uint8_t)ch_mask, (uint32_t)n_samples, OPTS_DEFAULT);
    prismlib_scan_read(dev, &status, n_samples, 60.0, buf, buf_size, &got);

    if (status & STATUS_BUFFER_OVERRUN)
        printf("WARNING: buffer overrun — 결과가 부정확할 수 있습니다.\n");

    wait_scan_stopped(dev);
    prismlib_scan_cleanup(dev);

    if (use_iepe)
        for (i = 0; i < n_ch; i++)
            prismlib_iepe_write(dev, (uint8_t)active[i], 0);

    int n_got = (int)got;
    if (n_got < n_samples) {
        printf("WARNING: %d 샘플 수신 (요청 %d) — FFT 크기를 %d으로 조정\n",
               n_got, n_samples, n_got);
        n_samples = n_got;
    }

    /* ── raw 데이터 저장 ─────────────────────────────────────────────────── */
    time_t t = time(NULL);
    struct tm *ti = localtime(&t);

    char raw_fname[64];
    strftime(raw_fname, sizeof(raw_fname), "raw_scan_%Y%m%d_%H%M%S.txt", ti);
    FILE *rfp = fopen(raw_fname, "w");
    if (rfp) {
        fprintf(rfp, "sample");
        for (int col = 0; col < n_ch; col++)
            fprintf(rfp, "\tch%d", active[col] + 1);
        fprintf(rfp, "\n");
        for (i = 0; i < n_samples; i++) {
            fprintf(rfp, "%d", i);
            for (int col = 0; col < n_ch; col++)
                fprintf(rfp, "\t%.6f", buf[(size_t)i * n_ch + col]);
            fprintf(rfp, "\n");
        }
        fclose(rfp);
        printf("Raw 데이터 저장: %s  (%d samples/ch × %d ch)\n",
               raw_fname, n_samples, n_ch);
    }

    /* ── FFT 분석 및 출력 ────────────────────────────────────────────────── */
    int bins = n_samples / 2 + 1;

    printf("\n");
    printf("%-8s  %13s  %12s  harmonics\n", "채널", "기본파 (Hz)", "진폭");
    printf("────────────────────────────────────────────────────────────────\n");

    double **ch_freq = NULL, **ch_amp = NULL;
    if (save_csv) {
        ch_freq = (double **)calloc(n_ch, sizeof(double *));
        ch_amp  = (double **)calloc(n_ch, sizeof(double *));
    }

    double pk_f[N_HARMONICS], pk_a[N_HARMONICS];
    int col;
    for (col = 0; col < n_ch; col++) {
        double *sig = (double *)malloc(n_samples * sizeof(double));
        if (!sig) break;
        for (i = 0; i < n_samples; i++)
            sig[i] = buf[(size_t)i * n_ch + col];

        double *freq = NULL;
        double *amp  = compute_spectrum(sig, n_samples, fs, &freq);
        free(sig);
        if (!amp) continue;

        int n_peaks = find_peaks(freq, amp, bins, pk_f, pk_a, N_HARMONICS);
        int h;

        printf("ch%-6d  %13.2f  %12.6f  ",
               active[col] + 1,
               n_peaks > 0 ? pk_f[0] : 0.0,
               n_peaks > 0 ? pk_a[0] : 0.0);
        for (h = 1; h < n_peaks; h++)
            printf("%.1fHz(%.4f)  ", pk_f[h], pk_a[h]);
        printf("\n");

        if (save_csv && ch_freq && ch_amp) {
            ch_freq[col] = freq;
            ch_amp[col]  = amp;
        } else {
            free(freq);
            free(amp);
        }
    }

    printf("────────────────────────────────────────────────────────────────\n");

    /* ── CSV 저장 ─────────────────────────────────────────────────────────── */
    if (save_csv && ch_freq && ch_amp) {
        char fname[64];
        strftime(fname, sizeof(fname), "fft_%Y%m%d_%H%M%S.csv", ti);

        FILE *fp = fopen(fname, "w");
        if (fp) {
            fprintf(fp, "frequency_hz");
            for (col = 0; col < n_ch; col++)
                fprintf(fp, ",ch%d_amplitude", active[col] + 1);
            fprintf(fp, "\n");
            for (i = 0; i < bins; i++) {
                fprintf(fp, "%.6g", ch_freq[0][i]);
                for (col = 0; col < n_ch; col++)
                    fprintf(fp, ",%.6g", ch_amp[col][i]);
                fprintf(fp, "\n");
            }
            fclose(fp);
            printf("\nCSV saved: %s\n", fname);
        }
        for (col = 0; col < n_ch; col++) {
            free(ch_freq[col]);
            free(ch_amp[col]);
        }
        free(ch_freq);
        free(ch_amp);
    }

    free(buf);
    prismlib_close(dev);
    return 0;
}

File: examples/python/fft_scan.py
#!/usr/bin/env python3
"""fft_scan.py — FFT 주파수 분석

유한 스캔으로 데이터를 수집하고 각 채널의 주파수 스펙트럼을 분석합니다.
피크 주파수와 고조파를 출력하고 결과를 CSV로 저장할 수 있습니다.

Usage:
    python fft_scan.py [ip] [port] [sample_rate] [channel_mask] [samples] [--csv]
    sample_rate:  0=64K 1=128K 2=170K 3=256K 4=512K  (기본 0=64K)
    channel_mask: 0x01~0x0F                           (기본 0x0F=전체)
    samples:      수집 샘플 수 (기본: sample_rate = 1초 분량)

Examples:
    python fft_scan.py 192.168.7.1 7777 1 0x01 128000
    python fft_scan.py 192.168.7.1 7777 0 0x0F 64000 --csv

Requires: numpy
"""
import sys
import time
import datetime
import numpy as np
from prismlib import prismlib, ScanOptions, ScanStatus

_SR_HZ   = [64_000, 128_000, 170_000, 256_000, 512_000]
_SR_NAME = ["64K",  "128K",  "170K",  "256K",  "512K" ]

N_HARMONICS = 5      # 기본파 포함 출력할 피크 수
MIN_FREQ_HZ = 5.0    # 이 주파수 미만은 피크 탐색 제외 (DC 제거)


# ── 인자 파싱 ──────────────────────────────────────────────────────────────────
ip      = sys.argv[1]         if len(sys.argv) > 1 else "192.168.7.1"
port    = int(sys.argv[2])    if len(sys.argv) > 2 else 7777
sr      = int(sys.argv[3])    if len(sys.argv) > 3 else 0
ch_mask = int(sys.argv[4], 0) if len(sys.argv) > 4 else 0x0F
save_csv = "--csv" in sys.argv

if not (0 <= sr <= 4):
    print("sample_rate: 0=64K 1=128K 2=170K 3=256K 4=512K", file=sys.stderr)
    sys.exit(1)
if not (0x01 <= ch_mask <= 0x0F):
    print("channel_mask: 0x01~0x0F", file=sys.stderr)
    sys.exit(1)

fs         = _SR_HZ[sr]
n_samples  = int(sys.argv[5]) if len(sys.argv) > 5 else fs   # 기본: 1초 분량
active_chs = [ch for ch in range(4) if ch_mask & (1 << ch)]
n_ch       = len(active_chs)


# ── FFT 유틸 ──────────────────────────────────────────────────────────────────

def analyze(signal, fs):
    """Hann 윈도우 적용, 단측 진폭 스펙트럼 반환: (freqs, amplitude)"""
    n        = len(signal)
    win      = np.hanning(n)
    gain     = win.mean()                   # 윈도우 진폭 보정 계수
    fft_vals = np.fft.rfft(signal * win)
    amp      = np.abs(fft_vals) / (n * gain)
    amp[1:-1] *= 2                          # 단측 스펙트럼 보정 (DC·Nyquist 제외)
    freqs    = np.fft.rfftfreq(n, d=1.0 / fs)
    return freqs, amp


def find_peaks(freqs, amp, n=N_HARMONICS):
    """기본파와 고조파 목록 반환: [(freq_hz, amplitude), ...]"""
    freq_res = float(freqs[1]) if len(freqs) > 1 else 1.0
    min_bin  = max(1, int(np.ceil(MIN_FREQ_HZ / freq_res)))

    peak_bin = min_bin + int(np.argmax(amp[min_bin:]))
    fund_hz  = float(freqs[peak_bin])

    result = []
    for h in range(1, n + 1):
        target_bin = int(round(fund_hz * h / freq_res))
        if target_bin >= len(freqs):
            break
        lo   = max(min_bin, target_bin - 2)
        hi   = min(len(freqs) - 1, target_bin + 2)
        best = lo + int(np.argmax(amp[lo : hi + 1]))
        result.append((float(freqs[best]), float(amp[best])))
    return result


# ── 스캔 및 분석 ───────────────────────────────────────────────────────────────
print(f"Connecting to {ip}:{port} ...")
prism = prismlib(ip, port)
prism.open()

try:
    prism.sampleRate_write(sr)
    print(f"Sample rate  : {_SR_NAME[sr]} ({fs:,} Hz)")
    print(f"Channels     : {'+'.join(f'ch{c+1}' for c in active_chs)}")
    print(f"Samples/ch   : {n_samples:,}  (주파수 분해능: {fs / n_samples:.3f} Hz)")
    ans = input("IEPE를 활성화하시겠습니까? [y/n]: ").strip().lower()
    use_iepe = ans in ("y", "yes")

    if use_iepe:
        for ch in active_chs:
            prism.iepe_write(ch, True)
        print("IEPE 안정화 대기 중 (2초)...", end="", flush=True)
        time.sleep(2.0)
        print(" 완료")

    print("Collecting ...", flush=True)

    prism.scan_start(ch_mask, n_samples, ScanOptions.DEFAULT)
    data_flat, status = prism.scan_read(n_samples, timeout=60.0)

    if status & ScanStatus.BUFFER_OVERRUN:
        print("WARNING: buffer overrun — 결과가 부정확할 수 있습니다.")

    while prism.scan_status()[0] & ScanStatus.RUNNING:
        time.sleep(0.005)
    prism.scan_cleanup()

    if use_iepe:
        for ch in active_chs:
            prism.iepe_write(ch, False)

    n_got = len(data_flat) // n_ch
    if n_got < n_samples:
        print(f"WARNING: {n_got} 샘플 수신 (요청 {n_samples}) — FFT 크기를 {n_got}으로 조정")
        n_samples = n_got

    # ── raw 데이터 저장 ─────────────────────────────────────────────────────────
    raw_fname = datetime.datetime.now().strftime("raw_scan_%Y%m%d_%H%M%S.txt")
    with open(raw_fname, "w") as f:
        f.write("sample\t" + "\t".join(f"ch{c+1}" for c in active_chs) + "\n")
        for i in range(n_samples):
            vals = data_flat[i * n_ch:(i + 1) * n_ch]
            f.write(f"{i}\t" + "\t".join(f"{v:.6f}" for v in vals) + "\n")
    print(f"Raw 데이터 저장: {raw_fname}  ({n_samples} samples/ch × {n_ch} ch)")

    # ── FFT 분석 및 출력 ────────────────────────────────────────────────────────
    print()
    print("─" * 64)
    print(f"{'채널':<6}  {'기본파 (Hz)':>13}  {'진폭':>12}  고조파")
    print("─" * 64)

    spectra = []
    for col, ch in enumerate(active_chs):
        signal     = np.array(data_flat[col::n_ch][:n_samples])
        freqs, amp = analyze(signal, fs)
        peaks      = find_peaks(freqs, amp)
        spectra.append((ch, freqs, amp, peaks))

        if not peaks:
            print(f"ch{ch+1:<4}  (유효한 피크 없음)")
            continue

        fund_hz, fund_amp = peaks[0]
        harmonics_str = "  ".join(
            f"{h_hz:.1f}Hz({h_amp:.4f})"
            for h_hz, h_amp in peaks[1:]
        )
        print(f"ch{ch+1:<4}  {fund_hz:>13.2f}  {fund_amp:>12.6f}  {harmonics_str}")

    print("─" * 64)

    # ── CSV 저장 ─────────────────────────────────────────────────────────────────
    if save_csv and spectra:
        import os
        timestamp = time.strftime("%Y%m%d_%H%M%S")
        fname     = f"fft_{timestamp}.csv"
        header    = "frequency_hz," + ",".join(f"ch{c+1}_amplitude" for c in active_chs)
        ref_freqs = spectra[0][1]
        cols      = [ref_freqs] + [amp for _, _, amp, _ in spectra]
        np.savetxt(fname, np.column_stack(cols),
                   delimiter=",", header=header, comments="", fmt="%.6g")
        print(f"\nCSV 저장: {os.path.abspath(fname)}")

finally:
    prism.close()
